src/pybind/mgr/cephadm/services/monitoring.py

```diff
@@ -357,7 +312,27 @@ def generate_config(
                 alerts = f.read()
             r['files']['/etc/prometheus/alerting/ceph_alerts.yml'] = alerts
 
-        return r, sorted(deps)
+        return r, sorted(self.calculate_deps())
+
+    def calculate_deps(self) -> List[str]:
+        deps = []  # type: List[str]
+        port = cast(int, self.mgr.get_module_option_ex(
+            'prometheus', 'server_port', self.DEFAULT_MGR_PROMETHEUS_PORT))
+        deps.append(str(port))
+        # add an explicit dependency on the active manager. This will force to
+        # re-deploy prometheus if the mgr has changed (due to a fail-over i.e).
+        deps.append(self.mgr.get_active_mgr().name())
+
+        node_exporter_cnt = len(self.mgr.cache.get_daemons_by_service('node-exporter'))
+        alertmgr_cnt = len(self.mgr.cache.get_daemons_by_service('alertmanager'))
+        haproxy_cnt = len(self.mgr.cache.get_daemons_by_type('ingress'))
+        if alertmgr_cnt > 0:
+            deps.append('alertmanager')
+        if haproxy_cnt > 0:
+            deps.append('ingress')
+        if node_exporter_cnt > 0:
+            deps.append('node-exporter')
```

Comment:
And same here. Why is this so redundant? It looks like the same code has to be added in 2 different places... 🤔

Reply:
You're right, it seems redundant because but right now cephadm `DaemonDescription` class has no notion of dependencies and they are calculated _manually_